Commands/PublishCommand.cs

```diff
@@ -65,52 +65,260 @@ public static int PublishFile(PublishOptions options, Config.Config config, stri
+
+                    var partRequest = new HttpRequestMessage(HttpMethod.Put, part.Url)
+                    {
+                        Content = new StreamContent(stream, part.Length)
+                    };
 
-            var response = client.Send(request);
-            if (response.StatusCode == System.Net.HttpStatusCode.OK)
+                    partRequest.Content.Headers.ContentMD5 = hash;
+                    
+                    // ReSharper disable once AccessToDisposedClosure
+                    // These tasks won't ever run past the client instance's lifetime
+                    using var response = await partClient.SendAsync(partRequest);
+
+                    try
+                    {
+                        response.EnsureSuccessStatusCode();
+                    }
+                    catch
+                    {
+                        Console.WriteLine(Red(await response.Content.ReadAsStringAsync()));
+                        throw;
+                    }
+
+                    return (true, new CompletedPartData()
+                    {
+                        ETag = response.Headers.ETag.Tag,
+                        PartNumber = part.PartNumber
+                    });
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine(Red("Exception occured while uploading file chunk:"));
+                    Console.WriteLine(Red(e.ToString()));
+
+                    return (false, null);
+                }
+            }
+
+            var uploadTasks = uploadData.UploadUrls.Select(UploadChunk).ToArray();
+            
+            static async Task<bool> ProgressBar(Task<(bool, CompletedPartData)>[] tasks)
             {
+                ushort spinIndex = 0;
+                string[] spinChars = { "|", "/", "-", "\\", "|", "/", "-", "\\" };
+                while (true)
+                {
+                    if (tasks.Any(x => x.IsCompleted && !x.Result.Item1))
+                    {
+                        Console.WriteLine();
+                        return false;
+                    }
+                    
+                    var completed = tasks.Count(static x => x.IsCompleted);
+                    
+                    Console.SetCursorPosition(0, Console.CursorTop);
+                    Console.Write(Green($"{completed}/{tasks.Length} chunks uploaded...{spinChars[spinIndex++ % spinChars.Length]}"));
+                    
+                    if (completed == tasks.Length)
+                    {
+                        Console.WriteLine();
+                        return true;
+                    }
+
+                    await Task.Delay(200);
+                }
+            }
 
+            if (!ProgressBar(uploadTasks).Result)
+            {
+                var abortRequest = new HttpRequestMessage(HttpMethod.Post, $"{config.PublishConfig.Repository}/api/experimental/usermedia/{uploadData.Metadata.UUID}/abort-upload/");

```

Comment:
I'd prefer if we kept the API url building behind dedicated methods, either in the config directly, or alternatively some other class dedicated for API path mapping (based on config)